library/core/src/num/dec2flt/parse.rs

```diff
@@ -1,121 +1,229 @@
-        Some(&b'e' | &b'E') => {
-            if integral.is_empty() {
-                return Invalid; // No digits before 'e'
-            }
+    // parse initial digits before dot
+    let mut mantissa = 0_u64;
+    let digits_start = s;
+    try_parse_digits(&mut s, &mut mantissa);
+    let mut n_digits = s.offset_from(&digits_start);
+
+    // handle dot with the following digits
+    let mut n_after_dot = 0;
+    let mut exponent = 0_i64;
+    let int_end = s;
+    if s.first_is(b'.') {
+        // SAFETY: s cannot be empty due to first_is
+        unsafe { s.step() };
+        let before = s;
+        try_parse_8digits(&mut s, &mut mantissa);
+        try_parse_digits(&mut s, &mut mantissa);
+        n_after_dot = s.offset_from(&before);
+        exponent = -n_after_dot as i64;
+    }
 
-            parse_exp(integral, b"", &s[1..])
+    n_digits += n_after_dot;
+    if n_digits == 0 {
+        return None;
+    }
+
+    // handle scientific format
+    let mut exp_number = 0_i64;
+    if s.first_is2(b'e', b'E') {
+        // SAFETY: s cannot be empty
+        unsafe {
+            s.step();
         }
-        Some(&b'.') => {
-            let (fractional, s) = eat_digits(&s[1..]);
-            if integral.is_empty() && fractional.is_empty() {
-                // We require at least a single digit before or after the point.
-                return Invalid;
-            }
+        // If None, we have no trailing digits after exponent, or an invalid float.
+        exp_number = parse_scientific(&mut s)?;
+        exponent += exp_number;
+    }
 
-            match s.first() {
-                None => Valid(Decimal::new(integral, fractional, 0)),
-                Some(&b'e' | &b'E') => parse_exp(integral, fractional, &s[1..]),
-                _ => Invalid, // Trailing junk after fractional part
-            }
+    // SAFETY: s must be larger than start and come from the same array
+    let len = s.offset_from(&start) as _;
+
+    // handle uncommon case with many digits
+    if n_digits <= 19 {
+        return Some((Number { exponent, mantissa, negative, many_digits: false }, len));
+    }
+
+    n_digits -= 19;
+    let mut many_digits = false;
+    let mut p = digits_start;
+    // SAFETY: p cannot be empty due to first_is2
+    unsafe {
+        while p.first_is2(b'0', b'.') {
+            // '0' = b'.' + 2
+            n_digits -= p.first_unchecked().saturating_sub(b'0' - 1) as isize;
+            p.step();
         }
-        _ => Invalid, // Trailing junk after first digit string
     }
-}
+    if n_digits > 0 {
+        // at this point we have more than 19 significant digits, let's try again
+        many_digits = true;
+        mantissa = 0;
+        let mut s = digits_start;
+        try_parse_19digits(&mut s, &mut mantissa);
+        exponent = if mantissa >= MIN_19DIGIT_INT {
+            // big int
+            int_end.offset_from(&s)
+        } else {
+            // SAFETY: the next byte must be present and be '.'
```

Comment:
How do you know this is true?

Reply:
I've updated the comment to specify the exact rationale:

```rust
        exponent = if mantissa >= MIN_19DIGIT_INT {
            // big int
            int_end.offset_from(&s)
        } else {
            // SAFETY: the next byte must be present and be '.'
            // We know this is true because we had more than 19
            // digits previously, so we overflowed a 64-bit integer,
            // but parsing only the integral digits produced less
            // than 19 digits. That means we must have a decimal
            // point, and at least 1 fractional digit.
            unsafe { s.step() };
            let before = s;
            try_parse_19digits(&mut s, &mut mantissa);
            -s.offset_from(&before)
        } as i64;
```

Reply:
Oh wow I would *not* have figured that out on my own :laughing: thanks for adding the comment!

Reply:
[Fixed](https://github.com/Alexhuszagh/rust/blob/870ed2163282b2e5fa99585e6c28c916e4861646/library/core/src/num/dec2flt/parse.rs#L179-L185) in the latest commit.